crates/red_knot_python_semantic/src/types.rs

```diff
@@ -3256,6 +3507,20 @@ impl<'db> Type<'db> {
             _ => KnownClass::Str.to_instance(db),
         }
     }
+
+    /// Reduce dynamic types in the type.
+    /// For example:
+    /// `T | Any | Unknown -> T | Any`
+    /// `T & Any & Unknown -> T & Any`
+    ///
+    /// This method is used to evaluate whether two types are gradually equal.
+    fn reduce_dynamic_types(&self, db: &'db dyn Db) -> Self {
+        match self {
+            Type::Union(union) => union.reduce_dynamic_types(db),
+            Type::Intersection(intersection) => intersection.reduce_dynamic_types(db),
+            _ => *self,
+        }
+    }
```

Comment:
I think perhaps we should just always do this eagerly in the union and intersection builders? Though there may be cases, particularly with Todo types, where this reduces debuggability. We can look into this separately.